File: solutions/go/collatz-conjecture/1/collatz_conjecture.go
package collatzconjecture

import "fmt"

func CollatzConjecture(n int) (int, error) {
	if (n < 1) {
        return -1, fmt.Errorf("%d is not a positive integer", n)
    }
    steps := 0
    for n > 1 {
        steps++
        if n & 1 == 0 {
            n >>= 1
        } else {
            n = (n<<1) + n + 1
        }
    }
    return steps, nil
}

File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start)
{
    if (start < 1)
    {
        return -1;
    }

    int result = 0;
    while (start != 1)
    {
        ++result;
        if (start%2)
        {
            start = 3*start + 1;
        }
        else
        {
            start = start/2;
        }
    }
    return result;
}
